### PR TITLE
layers: Sort stages for ValidateInterfaceBetweenStages

### PR DESCRIPTION
**Pull request by spencer-lunarg**
closes https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/8443

Before, we made the assumption that the order of `VkGraphicsPipelineCreateInfo::pStages` was how to get the consumer/producer stages for checking interface matching

We now sort the graphics stages correctly

### CODE REVIEW

File: tests/unit/shader_interface.cpp [hunk 1]
```diff
@@ -422,6 +422,39 @@ TEST_F(NegativeShaderInterface, VsFsTypeMismatch) {
...
+           color = vec4(x);
+        }
+    )glsl";
+
+    VkShaderObj vs(this, vsSource, VK_SHADER_STAGE_VERTEX_BIT);
+    VkShaderObj fs(this, fsSource, VK_SHADER_STAGE_FRAGMENT_BIT);
+
+    const auto set_info = [&](CreatePipelineHelper &helper) {
+        // Flipped here
+        helper.shader_stages_ = {fs.GetStageCreateInfo(), vs.GetStageCreateInfo()};
```
Comment (spencer-lunarg) [inline tests/unit/shader_interface.cpp:453]: This is an example where we had a clearly invalid 2 shaders, but by flipping them here, we were not detecting the error

**Review by arno-lunarg — APPROVED** [review]

File: layers/core_checks/cc_shader_interface.cpp [hunk 2]
```diff
@@ -901,19 +901,63 @@ bool CoreChecks::ValidateGraphicsPipelineShaderState(const vvl::Pipeline &pipeli
...
+                }
+            }
+            return not_found;
+        };
+        // Two graphic pipeline paths will be
+        // Vert -> (Tess) -> (Geom) -> [Fragment]
+        // (Task) -> Mesh -> [Fragment]
+        const size_t total_stages = 8;
+        // Pack both paths in, works because fragment are the last stage always
+        const VkShaderStageFlagBits stage_list[total_stages] = {VK_SHADER_STAGE_VERTEX_BIT,
```
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply layers/core_checks/cc_shader_interface.cpp:922]: I have never understood what value this provides

1. I don't want to list `8`, I need `total_stages` anyway below
2. what does this provide, why do I need to wrap this in a template when I don't use any of the template functions?

[the remark this answers is not in the document]
Comment (arno-lunarg) [reply layers/core_checks/cc_shader_interface.cpp:922]: total_stages is just ordered_stages.size() then
Only the methods you use will be instantiated, in this case `operator[]`. It provides bound checking in debug builds, without relying on ASAN to catch OOB errors

[the remark this answers is not in the document]
Comment (arno-lunarg) [reply layers/core_checks/cc_shader_interface.cpp:922]: I think you could even go `const std::array ordered_stages = { /*...*/ };` in this case

[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply layers/core_checks/cc_shader_interface.cpp:922]: ok, fair, will apply

[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply layers/core_checks/cc_shader_interface.cpp]: applied, agreed much cleaner looking now

**Pull request merged**